### DISCUSSION TITLE
単体テストの考え方/使い方 3.1.3〜3.1.8くらいまで

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#94] [General]
- 3.1 単体テストの構造の終わりくらい

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by yutafukuchi-1015** [comment]
## memo
### 3.1.4
- Arrange phase
  - オブジェクト・マザー・パターン
  - テストデータビルダー・パターン
- Act phase 
  - データの整合性が損なわれることを、`不変条件の侵害`という。
  - 不変条件の侵害が起こらないようにロジックをまとめることを、`カプセル化`という。
### 3.1.5
- Assert phase
  - `単体`とは1単位のコードでなく、1単位のふるまい。
  - よってAssert phaseは、複数の結果が生じることはあり得る
### 3.1.6
- 全phaseの後に、dbの初期化などの後処理は単体テストの場合、プロセス外依存とのやり取りは行わないため不要
### 3.1.7, 3.1.8
テスト対象に対してsutと宣言したり、何のフェーズなのかコメント、もしくは空白行を用いてコードを読みやすくする

**Comment by y-kaneda** [comment]
## 3.1 単体テストの構造
### 3.1.3 ~ 3.1.8

#### テストコードにif文を使うな
- 単純に見づらい
- 1つのケースで分岐が必要なほど多くのことを確認するな

#### 準備(Arrange)フェーズの大きさ
- 実行(Act)と確認(Assert)フェーズを合わせたより大きくなることはままある
- ただしプロダクトコードと同じようにメソッドを切り出したり共通化したりはするべし
  - オブジェクト・マザー(Object Mother)
  - テスト・データ・ビルダー(Test Data Builder)

#### 実行(Act)が複数行になるのはおかしい
- 1つのオペレーションに対して複数のコードを実行しないといけないのは危険
- ただ、確認したいケースの粒度とそれに対するコードの差な気もする
- 例えばインフラ層のDB操作のテストとかだったらアリなのでは
  - 「購入履歴ができたことを検証したい」で1ケース、「在庫が減っていることを検証したい」で1ケースならおかしくない気もする
    - usecase とかだったら1メソッドで両方できてないといけないのは分かる

#### 確認(Assert)で確認する項目
- 1単位のコードについてではなく1単位の振る舞いについて検証する

#### テストの後始末
- プロセス外依存とのやり取りを行わない前提なのが厳しくないか？DBに保存されてるか確認とかしたいはず

#### テスト対象とその依存の区別
- 最近だとGolandで自動生成してくれるテストの形式を使ってるからAAAの形かつ1メソッドだけ実行が分かりやすくなってるけど、もっと野生的に書く感じだと気を付けた方が良いかも

**Comment by xxManatoxx** [comment]
## 3.1.3 単体テストにおいて回避するべきこと:if文の使用

- if文はテストケース内であまりに多くの検証をしていることを示唆している
  - テストケースの条件分岐は可読性を悪くする

## 3.1.4 各フェーズのサイズはどのくらいが適切なのか？

- 準備フェーズが最も大きくなる
- 実行フェーズは基本は1メソッドになる
  - 1メソッドを超す場合はカプセル化の破綻を確認する
    - ただし、ビジネスロジックなどはこの考えが適用しやすいが、ユーティリティ、インフラに関するコードだと適切ではない場合があるので注意する

## 3.1.5 確認フェーズで確認する項目はどのくらいであればよいか？

- 単体テストの単体は1単位の振る舞いの為、複数の結果が生じる
  - 1つのテストケースで複数の確認があるのは自然な事
- ただし確認フェーズが大きくなりすぎることには注意が必要
  - 例えばオブジェクトを返すメソッドのテストの際にそのフィールドを1つずつ確認するのではなくオブジェクト同士の比較で確認ができるようにするべき

## 3.1.6 テストの後始末はどうするべきか？

- 単体テストでは基本的には後始末を必要としない
- 必要なのは統合テストの場合であり、その内容は第3部にて詳しく見る

## 3.1.7 テスト対象システムとその依存の区別

- テスト対象システムとその依存の違いが明確にわかる事が重要
  - 書籍ではテスト対象に対して「sut(System Under Test: SUT)」と命名している
- 確認点
  - Rspecではdescribeでテスト対象のクラスやメソッドを表現したりsubjectで表現するので命名で区別する必要はなさそう？

## 3.1.8 テストケースから各フェーズを示すコメントを取り除く

- AAAパターンを採用したケースについて
  - 準備フェーズや確認フェーズを書くのに空白行が不要ならコメントを取り除き空白行で各フェーズを区切るようにする
  - そうでない場合は各フェーズにフェーズを表すコメント（準備、実行、確認等）をつけるようにする

- 確認点
  - ここまでで確認フェーズや実行フェーズ等の書き方を見てきて以下感じた
    - 実行フェーズは基本1メソッド、書籍では区別できる名称をつけるので準備フェーズと実行フェーズはそれと空白行だけでも区別できるのでは？と感じた
      - 複数あった場合も適切に準備、実行、確認が分けられていれば問題ないようにも感じる
    - 確認フェーズは今自分が使用している言語や、書籍の言語などでは明確にassertやexpect等で表現しているのでこれも空白行で分けるだけで明確に区別できそうと感じた
    - 上記から書籍のコメントを残すことにあまりメリットを感じなかった

**Comment by myamamoto111** [comment]
## 3.1.3 〜 3.1.8

- 具体的な記述方法の解説で分かり易かった
  - if文を使わない
  - 実行（Act）フェーズのコードは１行
    - カプセル化
  - 確認（Assert）フェーズでの確認する項目は１単位の振る舞い
  - 単体テストで後始末はいらない
    - 後始末が必要なのは結合テスト
  - テスト対象システムに対して「sut」と名付ける
    - SUT: System Under Test
  - 各フェーズが明確に区別しやすくなっていること
    -  空白行
    - 各フェーズで空白行が必要な場合はコメント

**Comment by koushisa** [comment]
- 63p: 実行(Act)フェーズのコードが1行を超す場合は注意が必要
  - 実装分割の粒度を検証する指針として明確で良いと思った
  - 1つのシナリオを遂行するために複数のアクションが必要なのは設計上の問題を抱えているのではということに気づける

- 65p: 単体テストにおける「単体」とは、**1単位のコード** ではなく**1単位の振る舞い**のこと
  - 1単位の振る舞いによって複数の結果が生じることはありうる
  - 「振る舞いの粒度」 = 「ドメインの単位」
      - ドメインの単位はテストの目的や粒度(単体 or 結合 or etc...)によって変わる
      - 自分はこのコードのなにを検証したいのか？を意識して単体テストと結合テストの境界線を見極める
  - 状態遷移関数をnext = (prev * params)に分けて右辺の純粋関数を単体テストにする
  - 「振る舞いの粒度」を起点にDDD関連で思い出した個人的メモ
    - 大きなドメインはサブドメインに分割せよ
    - Domain Modeling Made Functionalだとworkflow->scenario->usecaseをパイプラインで繋ぐ


-  全体的にTDDの良さを感じた
    - 単体テストの目的を意識すると自然にインタフェース設計や実装分割の意識が生まれる
    - テストコードを書くことでこのような気づきが自然に生まれる
    - 自己完結するフィードバックを通して設計をインクリメンタルに改善できる

**Comment by nmurakami0** [comment]
3.1.3
- テストに分岐は設けないように

3.1.4
- Arrangeフェーズがもっとも大きくなる
    - ただしActとAssertの合計よりも大きくなるのならば、プライベートメソッドに切り出したり、その部分を作成する別のファクトリクラスを作ったほうが良い
        - オブジェクトマザーと呼ばれるパターンとテストデータビルダーと呼ばれるパターンがある
- Actフェーズは1行のコードで足りるはず
    - 足りない場合はテスト対象がきちんと設計されていない可能性が高い
    - ただ、ユーティリティやインフラに関するコードだとそうならない場合もある
- データの整合性が損なわれることは、不変条件(invariant)の侵害と呼ばれる
    - その対策がカプセル化になる

3.1.5
- 単体とは1単位のコードではなく、1つの振る舞いなので、複数の結果が生じることはありえる
    - なので、1つの確認しかすべきではないという認識は誤り
- ただしオブジェクトのフィールドを一つずつ比較するのではなくオブジェクトを同等と確認する手段を用意したほうが良い
- 備考
    - テストライブラリだと色々なmatcherが用意されている
        - https://rubydoc.info/github/rspec/rspec-expectations/RSpec/Matchers

3.1.6
- database_cleaner 的なのは後始末で行うイメージ

3.1.7
- テスト対象システムに対してsutという変数をつける
    - rspecの subject がそれに近そう
        - rspec的にはそれにさらに名前をつけるのが推奨されていた気がする
            - https://www.rubydoc.info/gems/rubocop-rspec/1.7.0/RuboCop/Cop/RSpec/NamedSubject

3.1.8
- 空白行またはコメントで各フェーズを表す
- テストライブラリによってはフェーズ分けをしやすい機能が用意されている印象
    - https://qiita.com/uchiko/items/d34c5d1298934252f58f
